zio-json/shared/src/main/scala/zio/json/JsonDecoder.scala

```diff
@@ -59,6 +59,26 @@ trait JsonDecoder[A] extends JsonDecoderPlatformSpecific[A] {
    */
   final def <*[B](that: => JsonDecoder[B]): JsonDecoder[A] = self.zipLeft(that)
 
+  final def both[B](that: => JsonDecoder[B]): JsonDecoder[(A, B)] =
+    bothWith(that)((a, b) => (a, b))
+
+  final def bothRight[B](that: => JsonDecoder[B]): JsonDecoder[B] =
+    bothWith(that)((_, b) => b)
+
+  final def bothLeft[B](that: => JsonDecoder[B]): JsonDecoder[A] =
+    bothWith(that)((a, _) => a)
+
+  final def bothWith[B, C](that: => JsonDecoder[B])(f: (A, B) => C): JsonDecoder[C] =
+    new JsonDecoder[C] {
+      override def unsafeDecode(trace: List[JsonError], in: RetractReader): C = {
+        val in2 = new WithRecordingReader(in, 64)
+        val a   = self.unsafeDecode(trace, in2)
+        in2.rewind()
```

Comment:
@plokhotnyuk I think there was some reason why rewind makes some issues or? Maybe there is a other way, if I am right about it having issues?

Reply:
`WithRecordingReader.rewind()` in `both*` combinators doesn't have issues itself, except possible harder to understand error messages and less efficiency.

But the strategy of re-parsing, especially if decoding error happens in almost the last place and subsequent decode starts from the beginning. The similar case of double parsing could happen with the discriminator field of ADT is placed in the end of the object and `WithRecordingReader.rewind()` is called after reaching the discriminator value to parse the case of ADT.

When this approach is used for nested or even recursive data structures it makes parsing vulnerable to DoS attacks, when attackers send specially selected inputs that reduces availability or increases scalability costs of systems greatly.

Reply:
I think that is worth a note on the doc but probably acceptable IMO. As @plokhotnyuk mentioned features like discriminators already present this attack surface. I don't think there's a better solution with the current API design.